connection_manager.go

```diff
@@ -0,0 +1,393 @@
+
+	// Mark ourselves as closed
+	cm.closed = true
+
+	// Close all watches.  The watch goroutine will take care of cleaning up
+	// connections and the watch map entry
+	for _, cw := range cm.watches {
+		cw.Close()
+	}
+
+	// Wait for watch goroutines to exit
+	cm.Unlock()
+	cm.wg.Wait()
+	return nil
+}
+
+// getExistingConn attmepts to return an existing open connection to the database cluster
+func (cm *connectionManager) getExistingConn(database, cluster string) (Connection, error) {
+	key := fmtConnKey(database, cluster)
+
+	cm.RLock()
+	defer cm.RUnlock()
+
+	// If we're closed, bail out
+	if cm.closed {
+		return nil, errClosed
+	}
+
+	// If there is already a connection, return it
+	if c := cm.conns[key]; c != nil {
+		return c, nil
+	}
+
+	// No connection, no error - try to create it
+	return nil, nil
+}
+
+// tryOpenConn tries to create a new connection to the database cluster
+func (cm *connectionManager) tryOpenConn(database, cluster string) (Connection, error) {
+	key := fmtConnKey(database, cluster)
+
+	cm.log.Infof("attempting to create connection to %s:%s", database, cluster)
+
+	cm.Lock()
+
+	// If someone closed us while we were upgrading the lock, bail out
+	if cm.closed {
+		return nil, errClosed
+	}
+
+	// If someone else created it while we were upgrading the lock, return the
+	// new connection
+	if c := cm.conns[key]; c != nil {
+		cm.Unlock()
+		return c, nil
+	}
+
+	// If someone is in the process of creating the connection, wait until they
+	// are done and try to retrieve it again
+	if p := cm.pending[key]; p != nil {
+		cm.Unlock()
+		<-p
+		return nil, nil
+	}
+
+	// Mark that we are creating the connection, so that we can release the lock
+	// and readers for other connections pass through while this connection is
+	// being established, without risking that we might create the connection
+	// more than once
+	p := make(chan struct{})
+	cm.pending[key] = p
+	cm.Unlock()
+
+	// Regardless of whether we succeed or fail, release anyone waiting
+	// for us to be done with the attempt to create
+	defer func() {
+		cm.Lock()
+		delete(cm.pending, key)
+		cm.Unlock()
+		close(p)
+	}()
+
+	// Get the current cluster configuration and watch for changes
+	cw, err := cm.provider.WatchCluster(database, cluster)
+	if err != nil {
+		cm.log.Errorf("failed to watch cluster config for %s:%s: %v", database, cluster, err)
+		return nil, err
+	}
+	<-cw.C()
+	cl := cw.Get()
+
+	// create the connection using the cluster configuration
+	conn, err := cm.openConn(cl)
+	if err != nil {
+		cm.log.Errorf("failed to open connection for %s:%s: %v", database, cluster, err)
+		cw.Close()
+		return nil, err
+	}
+
+	// register the connection and monitor the cluter for updates
+	if err := cm.registerConn(cl, cw, conn); err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// openConn opens a connection to the given cluster
+func (cm *connectionManager) openConn(cl Cluster) (Connection, error) {
+	d := cm.drivers[cl.Type().Name()]
```

Comment:
Worth merging line 247-260 with line 301-317 into a separate method that gets reused in both `openConn` and `reconfigureConn`?